runtime/platform.c

```diff
@@ -0,0 +1,276 @@
+
+#include <string.h>
+#include <unistd.h>
+#include <errno.h>
+#include <sys/time.h>
+#include "caml/platform.h"
+#include "caml/fail.h"
+#ifdef HAS_SYS_MMAN_H
+#include <sys/mman.h>
+#endif
+#ifdef _WIN32
+#include <windows.h>
+#endif
+
+/* Mutexes */
+
+void caml_plat_mutex_init(caml_plat_mutex * m)
+{
+  int rc;
+  pthread_mutexattr_t attr;
+  rc = pthread_mutexattr_init(&attr);
+  if (rc != 0) goto error1;
+  rc = pthread_mutexattr_settype(&attr, PTHREAD_MUTEX_ERRORCHECK);
+  if (rc != 0) goto error2;
+  rc = pthread_mutex_init(m, &attr);
+  // fall through
+error2:
+  pthread_mutexattr_destroy(&attr);
+error1:
+  check_err("mutex_init", rc);
+}
+
+void caml_plat_assert_locked(caml_plat_mutex* m)
+{
+#ifdef DEBUG
+  int r = pthread_mutex_trylock(m);
+  if (r == EBUSY) {
+    /* ok, it was locked */
+    return;
+  } else if (r == 0) {
+    caml_fatal_error("Required mutex not locked");
+  } else {
+    check_err("assert_locked", r);
+  }
+#endif
+}
+
+void caml_plat_assert_all_locks_unlocked()
+{
+#ifdef DEBUG
+  if (lockdepth) caml_fatal_error("Locks still locked at termination");
+#endif
+}
+
+void caml_plat_mutex_free(caml_plat_mutex* m)
+{
+  check_err("mutex_free", pthread_mutex_destroy(m));
+}
+
+static void caml_plat_cond_init_aux(caml_plat_cond *cond)
+{
+  pthread_condattr_t attr;
+  pthread_condattr_init(&attr);
+#if defined(_POSIX_TIMERS) && \
+    defined(_POSIX_MONOTONIC_CLOCK) && \
+    _POSIX_MONOTONIC_CLOCK != (-1)
+  pthread_condattr_setclock(&attr, CLOCK_MONOTONIC);
+#endif
+  pthread_cond_init(&cond->cond, &attr);
+}
+
+/* Condition variables */
+void caml_plat_cond_init(caml_plat_cond* cond, caml_plat_mutex* m)
+{
+  caml_plat_cond_init_aux(cond);
+  cond->mutex = m;
+}
+
+void caml_plat_wait(caml_plat_cond* cond)
+{
+  caml_plat_assert_locked(cond->mutex);
+  check_err("wait", pthread_cond_wait(&cond->cond, cond->mutex));
+}
+
+int caml_plat_timedwait(caml_plat_cond* cond, int64_t until)
+{
+  struct timespec t;
+  int err;
+  if (until < 0) {
+    /* until < 0 has definitely timed out, long ago.
+       letting the code below run risks feeding negative tv_nsec
+       to timedwait, since (-1 % 1000000000) = -1. */
+    return 1;
+  }
+  t.tv_sec  = until / 1000000000;
+  t.tv_nsec = until % 1000000000;
+  err = pthread_cond_timedwait(&cond->cond, cond->mutex, &t);
+  if (err == ETIMEDOUT) {
+    return 1;
+  } else {
+    return 0;
+  }
+}
+
+void caml_plat_broadcast(caml_plat_cond* cond)
+{
+  caml_plat_assert_locked(cond->mutex);
+  check_err("cond_broadcast", pthread_cond_broadcast(&cond->cond));
+}
+
+void caml_plat_signal(caml_plat_cond* cond)
+{
+  caml_plat_assert_locked(cond->mutex);
+  check_err("cond_signal", pthread_cond_signal(&cond->cond));
+}
+
+void caml_plat_cond_free(caml_plat_cond* cond)
+{
+  check_err("cond_free", pthread_cond_destroy(&cond->cond));
+  cond->mutex=0;
+}
+
+
+/* Memory management */
+
+#define Is_power_2(align) \
+  ((align) != 0 && ((align) & ((align) - 1)) == 0)
```

Comment:
Already defined in `<caml/misc.h>` with the name `Is_power_of_2`, except that this other definition misses the case `align = 0`.  Time to refactor?

Reply:
Fixed in e74f2d9